File: app/src/main/java/volovyk/guerrillamail/ui/MainViewModel.java
package volovyk.guerrillamail.ui;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.ViewModel;

import java.util.List;

import javax.inject.Inject;

import dagger.hilt.android.lifecycle.HiltViewModel;
import volovyk.guerrillamail.data.EmailRepository;
import volovyk.guerrillamail.data.SingleEvent;
import volovyk.guerrillamail.data.model.Email;

@HiltViewModel
public class MainViewModel extends ViewModel {
    private final EmailRepository emailRepository;
    private final LiveData<String> assignedEmail;
    private final LiveData<List<Email>> emails;
    private final LiveData<Boolean> refreshing;

    private final LiveData<SingleEvent<String>> errorLiveData;

    @Inject
    public MainViewModel(EmailRepository emailRepository) {
        this.emailRepository = emailRepository;
        this.assignedEmail = emailRepository.getAssignedEmail();
        this.emails = emailRepository.getEmails();
        this.refreshing = emailRepository.getRefreshing();
        this.errorLiveData = emailRepository.getErrorLiveData();
    }

    public LiveData<String> getAssignedEmail() {
        return assignedEmail;
    }

    public LiveData<List<Email>> getEmails() {
        return emails;
    }

    public LiveData<Boolean> getRefreshing(){ return refreshing;}

    public LiveData<SingleEvent<String>> getErrorLiveData() {
        return errorLiveData;
    }

    public void deleteEmail(Email email) {
        emailRepository.deleteEmail(email);
    }
}

File: app/src/main/java/volovyk/guerrillamail/ui/MainViewModel.kt
package volovyk.guerrillamail.ui

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.SharingStarted
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.map
import kotlinx.coroutines.flow.receiveAsFlow
import kotlinx.coroutines.flow.stateIn
import kotlinx.coroutines.launch
import timber.log.Timber
import volovyk.guerrillamail.R
import volovyk.guerrillamail.data.emails.EmailRepository
import volovyk.guerrillamail.data.emails.model.EmailRepositoryException
import javax.inject.Inject

data class UiState(
    val isLoading: Boolean = true,
    val isMainRemoteEmailDatabaseAvailable: Boolean = true
)

@HiltViewModel
class MainViewModel @Inject constructor(private val emailRepository: EmailRepository) :
    ViewModel() {

    init {
        Timber.d("init ${hashCode()}")
    }

    val uiState: StateFlow<UiState> = emailRepository.observeState()
        .map {
            UiState(
                isLoading = it.isLoading,
                isMainRemoteEmailDatabaseAvailable = it.isMainRemoteEmailDatabaseAvailable
            )
        }.stateIn(
            viewModelScope,
            SharingStarted.WhileSubscribed(5000),
            UiState()
        )

    val sideEffectFlow: Flow<SideEffect> = emailRepository.observeErrors()
        .receiveAsFlow()
        .map {
            when (it) {
                is EmailRepositoryException.EmailAddressAssignmentException ->
                    SideEffect.ShowToast(R.string.email_address_assignment_failure, it.message)

                is EmailRepositoryException.EmailFetchException ->
                    SideEffect.ShowToast(R.string.email_fetch_failure, it.message)
            }
        }

    fun retryConnectingToMainDatabase() {
        viewModelScope.launch {
            emailRepository.retryConnectingToMainDatabase()
        }
    }
}